src/libponyc/type/cap.c

```diff
@@ -912,49 +982,145 @@ bool cap_immutable_or_opaque(token_id cap)
+  ast_t* ephemeral = ast_sibling(cap);
+
+  token_id cap_token = ast_id(cap);
+  token_id eph_token = ast_id(cap);
+
+  if (mutation(&cap_token, &eph_token))
   {
-    case TK_ISO:
-      switch(cap)
+    type = ast_dup(type);
+    cap = cap_fetch(type);
+    ephemeral = ast_sibling(cap);
+
+    ast_setid(cap, cap_token);
+    ast_setid(ephemeral, eph_token);
+  }
+  return type;
+}
+
+ast_t* modified_cap(ast_t* type, cap_mutation* mutation)
+{
+  token_id id = ast_id(type);
+  switch(id)
+  {
+    case TK_UNIONTYPE:
+    case TK_ISECTTYPE:
+    case TK_TUPLETYPE:
+    {
+      ast_t* r_type = ast_from(type, ast_id(type));
+      ast_t* child = ast_child(type);
+
+      while(child != NULL)
       {
-        case TK_ISO:
-        case TK_VAL:
-        case TK_TAG:
-        case TK_CAP_SEND:
-        case TK_CAP_SHARE:
-          return true;
+        ast_append(r_type, modified_cap(child, mutation));
+        child = ast_sibling(child);
+      }
 
-        default: {}
+      return r_type;
+    }
+
+    case TK_NOMINAL:
+    case TK_TYPEPARAMREF:
+      return modified_cap_single(type, mutation);
+
+    case TK_ARROW:
+    {
+      // Alias just the right side. The left side is either 'this' or a type
+      // parameter, and stays the same.
+      AST_GET_CHILDREN(type, left, right);
+
+      ast_t* new_left = modified_cap(left, mutation);
+      ast_t* new_right = modified_cap(right, mutation);
+
+      if (new_left != left || new_right != right)
+      {
+        BUILD(r_type, type,
+          NODE(TK_ARROW,
+            TREE(new_left)
+            TREE(new_right)));
+        return r_type;
       }
-      break;
+      return type;
 
+    }
+
+    case TK_ISO:
     case TK_TRN:
-      switch(cap)
+    case TK_VAL:
+    case TK_REF:
+    case TK_BOX:
+    case TK_TAG:
+    {
+      token_id cap = id;
+      token_id eph = TK_NONE;
+
+      if (mutation(&cap, &eph))
       {
-        case TK_ISO:
-        case TK_VAL:
-        case TK_TAG:
-        case TK_CAP_SEND:
-        case TK_CAP_SHARE:
-          return true;
+        pony_assert(eph == TK_NONE);
+        type = ast_dup(type);
 
-        case TK_TRN:
-          return direction == WRITE;
-          
-        case TK_BOX:
-          return direction == EXTRACT;
-          
-        default: {}
+        ast_setid(type, cap);
       }
-      break;
+      return type;
+    }
 
-    case TK_REF:
+
+    case TK_THIS:
+    case TK_THISTYPE:
+
+    case TK_LAMBDATYPE:
+    case TK_BARELAMBDATYPE:
+    case TK_FUNTYPE:
+    case TK_INFERTYPE:
+    case TK_ERRORTYPE:
+    case TK_DONTCARETYPE:
+      return type;
+
+    default: {
+      ast_print(type, 80);
+      //ponyint_assert_fail(token_print(&id), __FILE__, __LINE__, __func__);
+      pony_assert(0);
+    }
+  }
+
+  pony_assert(0);
+  return NULL;
+}
+
+
+bool unisolated_mod(token_id* cap, token_id* eph)
+{
+  // Ephemeral caps have no isolation guarantees
+  // to maintain
+  if (*eph == TK_EPHEMERAL)
+    return cap;
```

Comment:
Did you mean to return `cap` (which is a `token_id*`)? The return type is `bool`, so I'm surprised the compiler is allowing it.

Maybe it's doing an implicit null check? If so, I'd prefer it to be explicit:
```suggestion
    return cap != NULL;
```

Reply:
Good catch, this should be returning false here.

Reply:
(This case does not actually occur currently)